src/extension.ts

```diff
@@ -233,46 +247,67 @@ class ProbeRSDebugAdapterServerDescriptorFactory implements vscode.DebugAdapterD
 			args.push("--vscode");
 
-			// Launch the debugger ... launch errors will be reported in `launchCallback`
+			// Launch the debugger ... launch errors will be reported in `onClose event`
 			logToConsole("INFO: Launching new server" + JSON.stringify(command) + " " + JSON.stringify(args) + " " + JSON.stringify(options));
-			var launchedDebugAdapter = child_process.execFile(
+			var launchedDebugAdapter = child_process.spawn(
 				command,
 				args,
 				options,
-				launchCallback,
 			);
 
 			// Capture stdout and stderr to ensure RUST_LOG can be redirected
-			debuggerReadySignature = "CONSOLE: Listening for requests on port " + debugServer[1];
+			var debuggerReadySignature = "CONSOLE: Listening for requests on port " + debugServer[1];
 			launchedDebugAdapter.stdout?.on('data', (data: string) => {
 				if (data.includes(debuggerReadySignature)) {
-					debuggerReady = true;
+					debuggerStatus = DebuggerStatus.running;
 				}
 				logToConsole(data);
 			});
 			launchedDebugAdapter.stderr?.on('data', (data: string) => {
-				logToConsole(data);
+				if (debuggerStatus === (DebuggerStatus.running as DebuggerStatus)) {
+					logToConsole("ERROR: " + data);
+				} else {
+					vscode.window.showErrorMessage("`probe-rs-debugger` error: " + data);
+				}
+			});
+			launchedDebugAdapter.on('close', (code: number | null, signal: string | null) => {
+				if (debuggerStatus !== (DebuggerStatus.failed as DebuggerStatus)) {
+					handleExit(code, signal);
+				}
+			});
+			launchedDebugAdapter.on('error', (err: Error) => {
+				if (debuggerStatus !== (DebuggerStatus.failed as DebuggerStatus)) {
+					debuggerStatus = DebuggerStatus.failed;
+					vscode.window.showErrorMessage("`probe-rs-debugger` process encountered an error: " + JSON.stringify(err));
+					launchedDebugAdapter.kill();
+				}
 			});
-
 			// Wait to make sure probe-rs-debugger startup completed, and is ready to accept connections.
 			var msRetrySleep = 250;
 			var numRetries = 5000 / msRetrySleep;
-			while (!debuggerReady) {
+			while (debuggerStatus === DebuggerStatus.starting) {
 				await new Promise<void>((resolve) => setTimeout(resolve, msRetrySleep));
 				if (numRetries > 0) {
 					numRetries--;
 				} else {
-					launchedDebugAdapter.kill();
+					debuggerStatus = DebuggerStatus.failed;
 					logToConsole("ERROR: Timeout waiting for probe-rs-debugger to launch");
 					vscode.window.showErrorMessage("Timeout waiting for probe-rs-debugger to launch");
-					return undefined;
+					break;
 				}
 			}
-			await new Promise<void>((resolve) => setTimeout(resolve, 500)); // Wait for a fraction of a second more, to allow TCP/IP port to initialize in probe-rs-debugger
+
+			if (debuggerStatus === (DebuggerStatus.running as DebuggerStatus)) {
+				await new Promise<void>((resolve) => setTimeout(resolve, 500)); // Wait for a fraction of a second more, to allow TCP/IP port to initialize in probe-rs-debugger
+			}
```

Comment:
Not directly related to this PR, but what is the reason to use a TCP connection by default?

It seems easier to me to just use stdout to communicate, which requires less setup.

Reply:
This is how I originally had it. I discovered it became unstable (mostly lost requests/responses), which were not reproducible over TCP. 

Once I made the changeover, I also found it opened the opportunity to pipe RUST_LOG output to the user's VSCode console, so I figured it was worth the extra effort at extension development time.

What do you mean "requires less setup"?

Reply:
With stdout, I would assume that that we don't need to wait for any setup, and it would also not be required to find a suitable port to communicate.

The logs are sent to stderr AFAIk, so catching them should still be possible.

Reply:
Yep, like I said, that is how I had it working. I implemented the TCP by default based on some posts I found where other extension developers saw similar instability with the STDIO version.

Reply:
I'm a bit surprised that STDIO is unstable, but have no reason to doubt it :)

Could you add a comment to explain this? Just to avoid similar questions in the future. Thanks!